### DISCUSSION TITLE
boilerplate monopoly com port?

### DISCUSSION THREAD

**Discussion by dodo1004** [cncjs/cncjs#781] [Q&A] [answered]
Hi.
I am making new pendant.
so I am checking out How pendants are running.
When boilerplate pendants connect socket io and  serial port, 
cncjs instants can't access serial port?
In web-browser, Cncjs instants share com port.
but when I connect pendant, they can't open serial port.
and reverse is the same.
Normally, Pendant should run with main program.'
I do not know It is intend to be made or Just a bug.
I didn't see what is normal state.
Anybody tell me about it. then I am really appreciated.

**Comment by MitchBradley** [comment, marked as answer]
Serial ports cannot be shared between programs.  For a pendant to send to a port, it must communicate with the CNCjs app and ask it to send to the port.  Study the existing pendants to see how they work.

**Reply by dodo1004** [reply]
In Real Serial Device, They can't be shared.
but, It is a Server that Every CNCjs apps connect the server and Just only one is really connected at Serial GRBL devcie but others are virtually connected That is adjusted by CNCjs server or first instance.
 ( When I opened a CNCjs app and turned off the server, the app was running well)
Right? I am a C++ programmer not node js , Reading CNCjs source is  overdoing for me so I need to learn abstraction layer that is any Instance what they are doing and where they are connected. 
I need just right concepts and I use "socket io client  c++"
but when I learn deeper socket io c++, It compel to learn and read node js sources

Finally Can I ask again to be sure ( that makes me confuse)
1. when a pendent is used, CNCjs app have to close socket? (because pendant need to connect socket?)
2. CNCjs apps share grbl port virtually, but current pendant example can't be share like CNCjs app?
3. if Really pendant example need Real Grbl com port device  ( I think that io.connect() method could be made virtually in cncjs )
       then Do we have any way to connect grbl virtually like other CNCjs app?

Thank you for your answer, I really appreciated.

----------------------------------------------------------------------------
I am working at Raspberry pi 4
and when I running cncjs's pendants , dependent libraries version is so old that was not running.
I got running after changing package.json file(just only lib version up)
so it is changed and I do not know pure running state.
In this state, When I run boilerplate pendant, It request serial comport name, If com port is not using 
then pendant can connect the port but if comport is using by other apps then It stop running and abnormally exit with debugging messages. even I do not know It is running normally like that or not.
like your answer "it must communicate with the CNCjs app and ask it to send to the port."
If it is occupied by other process then pendant app should give me another way to communicate with grbl port virtually or with any asking method. unfortunately It was just killed.
It is my confusion Its state is intended or some bug has or only for independent pendant.

**Reply by dodo1004** [reply]
The Problem is solved
In my raspberry pi
CNCJs app 's default ip was 127.0.0.1 
CNCjs server's default ip was 127.0.1.1
So I lost device connection.
In the same server, they share a port.

but I do not know how app is running without server.(maybe default server exist anywhere)

> boilerplate monopoly com port? 
> NO, They share a port in the same server
